--- a/fs/namespace.c
+++ b/fs/namespace.c
@@ -76,6 +76,28 @@
 
 /* Spinlock for 'struct mount' member changes */
 static DEFINE_SPINLOCK(mount_lock);
+
+int path_umount(struct path *path, int flags)
+{
+	struct mount *mnt = real_mount(path->mnt);
+	int ret;
+
+	if (flags & ~(MNT_FORCE | MNT_DETACH | MNT_EXPIRE | UMOUNT_NOFOLLOW))
+		return -EINVAL;
+	if (!may_mount())
+		return -EPERM;
+	if (path->dentry != path->mnt->mnt_root)
+		return -EINVAL;
+
+	ret = do_umount(mnt, flags);
+
+	dput(path->dentry);
+	mntput_no_expire(mnt);
+	return ret;
+}
+EXPORT_SYMBOL(path_umount);
 
 static inline void lock_mount_hash(void)
 {
--- a/include/linux/fs.h
+++ b/include/linux/fs.h
@@ -2448,6 +2448,7 @@
 extern int do_remount(struct path *, int, int, void *);
 #endif
 extern int do_umount(struct vfsmount *, int);
+extern int path_umount(struct path *, int);
 extern int do_mount(const char *, const char __user *, const char *, unsigned long, void *);
 extern struct vfsmount *collect_mounts(const struct path *);
 extern void drop_collected_mounts(struct vfsmount *);
